File: PythonFastApi/app/services/user_service.py
from app.repositories.user_repository import UserRepository
from app.models import User

class UserService:
    def __init__(self, user_repository: UserRepository):
        self.user_repository = user_repository

    async def create_user(self, name: str, email: str) -> User:
        existing_user = await self.user_repository.get_user_by_email(email)
        if existing_user:
            return None # Email already taken
        return await self.user_repository.create_user(name, email)

    async def get_all_users(self) -> list[User]:
        return await self.user_repository.get_all_users()

    async def get_user_by_id(self, user_id: int) -> User | None:
        return await self.user_repository.get_user_by_id(user_id)

    async def update_user(self, user_id: int, name: str | None, email: str | None) -> User | None:
        user = await self.user_repository.get_user_by_id(user_id)
        if not user:
            return None

        if email:
             existing_user = await self.user_repository.get_user_by_email(email)
             if existing_user and existing_user.id != user_id:
                 return None # Email taken by another user

        return await self.user_repository.update_user(user, name, email)

    async def delete_user(self, user_id: int) -> bool:
        user = await self.user_repository.get_user_by_id(user_id)
        if not user:
            return False
        await self.user_repository.delete_user(user)
        return True

File: JavaGraalVM/src/main/java/com/magnetaryggdrasil/service/UserService.java
package com.magnetaryggdrasil.service;

import com.magnetaryggdrasil.model.User;
import com.magnetaryggdrasil.repository.UserRepository;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import java.util.List;

@ApplicationScoped
public class UserService {

    @Inject
    UserRepository userRepository;

    public List<User> getAllUsers() {
        return userRepository.listAll();
    }

    public User getUserById(Long id) {
        return userRepository.findById(id);
    }

    @Transactional
    public User createUser(User user) {
        if (userRepository.findByEmail(user.email) != null) {
            throw new IllegalArgumentException("Email already exists");
        }
        userRepository.persist(user);
        return user;
    }

    @Transactional
    public User updateUser(Long id, User userDetails) {
        User user = userRepository.findById(id);
        if (user == null) {
            return null;
        }
        if (userDetails.name != null) {
            user.name = userDetails.name;
        }
        if (userDetails.email != null) {
             if (!user.email.equals(userDetails.email) && userRepository.findByEmail(userDetails.email) != null) {
                throw new IllegalArgumentException("Email already exists");
            }
            user.email = userDetails.email;
        }
        return user;
    }

    @Transactional
    public boolean deleteUser(Long id) {
        return userRepository.deleteById(id);
    }
}
